File: public/display/GrandPrix.js
// Grand Prix series engine — the pure cup-scoring core behind "pick a Cup" mode.
// A CupSeries walks one cup's tracks in CUPS order, banks points per race
// (POINTS_BY_RANK by finish rank, DNF = 0) and answers the intermission /
// podium standings. Display-side only, but kept dependency-free, clock-free and
// RNG-injected (like engine/Game.js) so node:test drives it directly and the
// portable-purity gate scans it: all identity/meta comes in through applyRace,
// randomness comes in through makeShuffleBag's rng.

// Points by finish rank (1st..4th). The field is always FIELD_SIZE = 4 (humans
// + AI top-up), so this is the whole table; DNF earns nothing regardless of rank.
export const POINTS_BY_RANK = [9, 6, 3, 1];

// Endless track draw with no repeats until every id has been seen: deal from a
// shuffled copy, reshuffle when empty, and never let the first draw of a fresh
// bag repeat the last draw of the old one. rng is injected (page Math.random in
// the display, a seeded LCG in tests) — () => [0,1).
export function makeShuffleBag(ids, rng) {
  const all = [...ids];
  let bag = [];
  let last = null;
  const refill = () => {
    bag = [...all];
    for (let i = bag.length - 1; i > 0; i--) { // Fisher–Yates
      const j = Math.floor(rng() * (i + 1));
      [bag[i], bag[j]] = [bag[j], bag[i]];
    }
    // draw() pops from the END — swap a boundary repeat to the far end of the bag
    if (all.length > 1 && bag[bag.length - 1] === last) {
      [bag[0], bag[bag.length - 1]] = [bag[bag.length - 1], bag[0]];
    }
  };
  return {
    draw() {
      if (!bag.length) refill();
      last = bag.pop();
      return last;
    }
  };
}

// One cup's 4-race series. Participants are keyed by playerId (peerIndex for
// humans, 'ai-N' for bots — stable while the roster holds). Everyone ever seen
// stays on the board: a dropped player keeps their points and simply stops
// gaining; a late joiner scores from the race they first appear in.
export class CupSeries {
  constructor(cup) {
    this.cup = cup;              // a CUPS entry: {id, name, tracks: [4 ids]}
    this.raceIndex = 0;          // 0-based race currently being raced / just raced
    this._lastApplied = -1;      // raceIndex of the last applyRace (for tie-breaks)
    this._done = false;          // final race's results are in → podium time
    this._meta = new Map();      // playerId → {name, colorIndex, ai, points, gained, lastRank, lastRaceIndex}
  }

  get raceCount() { return this.cup.tracks.length; }
  get currentTrackId() { return this.cup.tracks[this.raceIndex]; }
  get nextTrackId() {
    return this.raceIndex + 1 < this.raceCount ? this.cup.tracks[this.raceIndex + 1] : null;
  }
  get finished() { return this._done; }

  // Bank one race. results = Game.getResults().results ({playerId, rank,
  // finished, time}); field = the display's currentField ({peerIndex, name,
  // colorIndex, ai, ...}) — the only side that can name/colour the AI seats.
  applyRace(results, field) {
    const seats = new Map(field.map((p) => [p.peerIndex, p]));
    for (const m of this._meta.values()) m.gained = 0; // absentees show "+0" this round
    for (const res of results) {
      const seat = seats.get(res.playerId) || {};
      let m = this._meta.get(res.playerId);
      if (!m) {
        m = { name: seat.name, colorIndex: seat.colorIndex, ai: !!seat.ai, points: 0, gained: 0, lastRank: null, lastRaceIndex: -1 };
        this._meta.set(res.playerId, m);
      }
      m.gained = res.finished ? (POINTS_BY_RANK[res.rank - 1] || 0) : 0;
      m.points += m.gained;
      m.lastRank = res.rank;
      m.lastRaceIndex = this.raceIndex;
    }
    this._lastApplied = this.raceIndex;
    if (this.raceIndex >= this.raceCount - 1) this._done = true;
  }

  // Move to the next race (the intermission's "Next race ▸"). No-op past the end.
  advance() {
    if (this.raceIndex + 1 < this.raceCount) this.raceIndex++;
  }

  // Cup order: points, then placement in the LATEST race (whoever beat whom just
  // now wins the tie; sitting that race out loses it), then first-seen order —
  // sort() is stable, so double-ties never shuffle between broadcasts.
  standings() {
    const rows = [...this._meta.entries()].map(([playerId, m]) => ({
      playerId, name: m.name, colorIndex: m.colorIndex, ai: m.ai,
      points: m.points, gained: m.gained, lastRank: m.lastRank
    }));
    const latest = (id) => {
      const m = this._meta.get(id);
      return m.lastRaceIndex === this._lastApplied ? m.lastRank : Infinity;
    };
    return rows.sort((a, b) => b.points - a.points || latest(a.playerId) - latest(b.playerId));
  }

  // A reconnect lands the same human on a new peerIndex (see rekeyCarPlayer);
  // carry their banked cup over. Rebuilt in place so first-seen order (the
  // stable tie-break) survives the rename.
  rekey(oldId, newId) {
    if (oldId === newId || !this._meta.has(oldId)) return;
    const moved = new Map();
    for (const [id, m] of this._meta) moved.set(id === oldId ? newId : id, m);
    this._meta = moved;
  }
}

File: native/libttp-sim/ttp/grand_prix.cc
#include "ttp/grand_prix.h"

#include <algorithm>
#include <cmath>
#include <limits>

namespace ttp {

const int POINTS_BY_RANK[4] = {9, 6, 3, 1};

ShuffleBag::ShuffleBag(std::vector<std::string> ids, std::function<double()> rng)
    : all_(std::move(ids)), rng_(std::move(rng)) {}

void ShuffleBag::refill() {
  bag_ = all_;
  for (int i = (int)bag_.size() - 1; i > 0; i--) {  // Fisher-Yates
    int j = (int)std::floor(rng_() * (i + 1));
    std::swap(bag_[i], bag_[j]);
  }
  // draw() pops from the END — swap a boundary repeat to the far end of the bag.
  if (all_.size() > 1 && hasLast_ && bag_.back() == last_) std::swap(bag_.front(), bag_.back());
}

std::string ShuffleBag::draw() {
  if (bag_.empty()) refill();
  last_ = bag_.back(); hasLast_ = true;
  bag_.pop_back();
  return last_;
}

CupSeries::CupSeries(const GpCup& cup, std::function<std::string()> drawNext)
    : cup_{cup.id, cup.name, cup.tracks}, drawNext_(std::move(drawNext)) {}

CupSeries::Meta* CupSeries::findMeta(const Id& id) {
  for (auto& kv : meta_) if (kv.first == id) return &kv.second;
  return nullptr;
}
const CupSeries::Meta* CupSeries::findMeta(const Id& id) const {
  for (const auto& kv : meta_) if (kv.first == id) return &kv.second;
  return nullptr;
}

void CupSeries::applyRace(const std::vector<GpResult>& results, const std::vector<GpFieldEntry>& field) {
  for (auto& kv : meta_) kv.second.gained = 0;  // absentees show "+0"
  for (const GpResult& res : results) {
    const GpFieldEntry* seat = nullptr;
    for (const auto& p : field) if (p.peerIndex == res.playerId) { seat = &p; break; }
    Meta* m = findMeta(res.playerId);
    if (!m) {
      Meta nm;
      nm.seatNull = (seat == nullptr);
      if (seat) { nm.name = seat->name; nm.colorIndex = seat->colorIndex; nm.ai = seat->ai; }
      meta_.emplace_back(res.playerId, nm);
      m = &meta_.back().second;
    }
    m->gained = res.finished ? (res.rank >= 1 && res.rank <= 4 ? POINTS_BY_RANK[res.rank - 1] : 0) : 0;
    m->points += m->gained;
    m->lastRankNull = false;
    m->lastRank = res.rank;
    m->lastRaceIndex = raceIndex_;
  }
  lastApplied_ = raceIndex_;
  if (raceIndex_ >= raceCount() - 1) {
    if (drawNext_) cup_.tracks.push_back(drawNext_());
    else done_ = true;
  }
}

void CupSeries::advance() {
  if (raceIndex_ + 1 < raceCount()) raceIndex_++;
}

std::vector<GpStanding> CupSeries::standings() const {
  std::vector<GpStanding> rows;
  for (const auto& kv : meta_) {
    const Meta& m = kv.second;
    rows.push_back({kv.first, m.name, m.colorIndex, m.ai, m.points, m.gained,
                    m.lastRankNull, m.lastRank, m.seatNull});
  }
  const double INF = std::numeric_limits<double>::infinity();
  auto latest = [&](const Id& id) -> double {
    const Meta* m = findMeta(id);
    return (m && m->lastRaceIndex == lastApplied_) ? (double)m->lastRank : INF;
  };
  std::stable_sort(rows.begin(), rows.end(), [&](const GpStanding& a, const GpStanding& b) {
    if (a.points != b.points) return a.points > b.points;  // b.points - a.points
    return latest(a.playerId) < latest(b.playerId);
  });
  return rows;
}

std::vector<Id> CupSeries::lastRaceOrder() const {
  std::vector<std::pair<int, Id>> ranked;
  for (const auto& kv : meta_) {
    const Meta& m = kv.second;
    if (!m.lastRankNull && m.lastRaceIndex == lastApplied_) ranked.push_back({m.lastRank, kv.first});
  }
  std::stable_sort(ranked.begin(), ranked.end(),
                   [](const auto& a, const auto& b) { return a.first < b.first; });
  std::vector<Id> out;
  out.reserve(ranked.size());
  for (const auto& r : ranked) out.push_back(r.second);
  return out;
}

void CupSeries::rekey(const Id& oldId, const Id& newId) {
  if (oldId == newId || !findMeta(oldId)) return;
  for (auto& kv : meta_) if (kv.first == oldId) kv.first = newId;
}

}  // namespace ttp
